### PR TITLE
[eval] fix event loop

### LINKED ISSUE #10682: Mainloop terminating unintentionally
This issue is from Heaps originally, and as requested, is moved here.

Original issues https://github.com/HeapsIO/heaps/issues/1068

---- copied from heaps issue
minimum code https://github.com/ZwodahS/heaps-sound-bug

OS : Mac
heaps version: 57c1bd34779193636562621faca5a875b2113923
haxe: 4.2.4

SoundManager set isBlocking to false.
https://github.com/HeapsIO/heaps/blob/master/hxd/snd/Manager.hx#L151

Main loop cancel the event and only adds it back when there is a blocking event.
https://github.com/HaxeFoundation/haxe/blob/development/std/haxe/MainLoop.hx#L81-L89
https://github.com/HaxeFoundation/haxe/blob/development/std/haxe/MainLoop.hx#L124

This is fine when live update is set to true `hxd.res.Resource.LIVE_UPDATE = true;` (suspect that the live update created a blocking main event) or when you have mainEvent elsewhere.

with that, there is 2 solutions that I found that works
https://github.com/ZwodahS/heaps-sound-bug/blob/main/App.hx#L4-L5

However, I am curious as to what is the best way to handle this, and if this is considered a bug.
----

### LINKED ISSUE #10832: MainLoop issues on HEAD in Eval
Just effects eval

commit b2f0e1bc4f7b02e529ee1199028b555452d9f1ca

```haxe
import haxe.MainLoop;

class Main {
	static function main() {
		trace('main');
		var event = null;
		var done 	= false;
		event = MainLoop.add(
			() -> {
				if(!done){
					done = true;
					trace('ok');
				}
				if(event !=null){
					trace("stop");
					event.stop();
				}
			}
		);
	}
}
```

EDIT: repeat of `var event`.

### PR DESCRIPTION
**Pull request by kLabz**
This PR is not checking output because it actually varies a bit (mostly the ABCDE traces) on all targets (including the ones I am not modifying here) and platforms, due to the nature of the test mixing timers, addThread and MainLoop.add.

This PR also revives tests previously disabled, which should close #10682, and add a new test to close #10832

## Test result on eval 4.2.5 (locally)

```
Main.hx:7: 0
Main.hx:48: A
Main.hx:7: 1
Main.hx:48: B
Main.hx:7: 2
Main.hx:48: C
Main.hx:7: 3
Main.hx:48: D
Main.hx:7: 4
Main.hx:48: E
Main.hx:7: 5
Main.hx:7: 6
Main.hx:7: 7
Main.hx:7: 8
Main.hx:7: 9
Main.hx:10: false
Main.hx:18: BEFORE1
Main.hx:35: BEFORE2
Main.hx:26: T0
Main.hx:26: T1
Main.hx:26: T2
Main.hx:26: T3
Main.hx:26: T4
```

Other 4.2.x gave the same result but hanged at the end

## Test result on recent nightlies

```
Main.hx:18: BEFORE1
Main.hx:35: BEFORE2
Main.hx:26: T0
Main.hx:26: T1
Main.hx:26: T2
Main.hx:26: T3
Main.hx:26: T4
```

## Test result on all targets with this PR

Note that the point at which ABCDE traces start can vary a bit

```
Main.hx:18: BEFORE1
Main.hx:35: BEFORE2
Main.hx:26: T0
Main.hx:48: A
Main.hx:7: 0
Main.hx:48: B
Main.hx:7: 1
Main.hx:48: C
Main.hx:7: 2
Main.hx:48: D
Main.hx:7: 3
Main.hx:48: E
Main.hx:7: 4
Main.hx:7: 5
Main.hx:7: 6
Main.hx:7: 7
Main.hx:7: 8
Main.hx:7: 9
Main.hx:10: false
Main.hx:26: T1
Main.hx:26: T2
Main.hx:26: T3
Main.hx:26: T4
```

`Main.hx:7: 0` will also likely run before the timers on eval; I used to fight it but doing so breaks #10832, and I'm honestly not sure what actual problem this brings.

### CODE REVIEW

File: std/eval/_std/sys/thread/EventLoop.hx [hunk 1]
```diff
@@ -96,45 +128,79 @@ class EventLoop {
...
-		//TODO: throw if loop is already running
+		if (started) throw "Event loop already started";
+		started = true;
 		consumePending();
-		(handle:Loop).run(DEFAULT);
+		handle.run(DEFAULT);
 	}
 
 	function consumePending(?_:Async):Void {
+		mutex.acquire();
```
Comment (Simn) [inline std/eval/_std/sys/thread/EventLoop.hx:191]: Does this actually do anything here? The only possible side-effect is that `pending` gets reassigned after reading it, but in that case just assigning `[]` to it doesn't seem to be correct either.

Comment (Simn) [reply std/eval/_std/sys/thread/EventLoop.hx:191]: Oh wait, I'm answering my own question because that's precisely the case this handles.

Comment (kLabz) [reply std/eval/_std/sys/thread/EventLoop.hx:191]: Indeed. I guess it's not needed in `repeat` anymore either.

File: std/eval/_std/sys/thread/EventLoop.hx [hunk 2]
```diff
@@ -96,45 +128,79 @@ class EventLoop {
...
+		Returns `true` if more events are expected.
+		Returns `false` if no more events expected.
+
+		Depending on a target platform this method may be non-reentrant. It must
+		not be called from event callbacks.
+	**/
 	public function wait(?timeout:Float):Bool {
-		//TODO: throw if loop is already running
-		if(timeout == null) {
+		if (started) throw "Event loop already started";
```
Comment (Simn) [inline std/eval/_std/sys/thread/EventLoop.hx:161]: At some point we should start throwing something other than raw strings... There might be a use-case for try-catching a `wait` call and specifically handling this exception, as opposed to some exception that comes from further down the call-stack.

On the other hand, one could just check the `started` state manually.

Comment (kLabz) [reply std/eval/_std/sys/thread/EventLoop.hx:161]: tbh I'm not really sure what this wait is supposed to be used for, as it was broken before this PR (no timeout provided = crash, timeout provided = timeout ignored)

Comment (Simn) [reply std/eval/_std/sys/thread/EventLoop.hx:161]: Isn't this what you periodically call when you want to handle events but manage the main loop yourself?

Comment (kLabz) [reply std/eval/_std/sys/thread/EventLoop.hx:161]: I suppose; I don't do that kind of things myself :sweat_smile:

**Pull request merged**